core/src/main/scala/org/broadinstitute/dsde/rawls/bucketMigration/BucketMigrationService.scala

```diff
@@ -27,6 +34,77 @@ class BucketMigrationService(val dataSource: SlickDataSource, val samDAO: SamDAO
     with WorkspaceSupport
     with LazyLogging {
 
+  def getEligibleOrMigratingWorkspaces: Future[Map[String, Option[MultiregionalBucketMigrationProgress]]] = {
+    def getOwnedWorkspaces: Future[Seq[Workspace]] = for {
+      userWorkspaces <- samDAO.listUserResources(SamResourceTypeNames.workspace, ctx)
+      workspaceIds = userWorkspaces
+        .filter(_.hasRole(SamWorkspaceRoles.owner))
+        .map(workspaceResource => Try(UUID.fromString(workspaceResource.resourceId)))
+        .flatMap(_.toOption)
+      workspaces <- dataSource.inTransaction(_.workspaceQuery.listV2WorkspacesByIds(workspaceIds))
+    } yield workspaces
+
+    def getEligibleUnmigratedWorkspaces(
+      workspaces: Seq[Workspace]
+    ): Future[Seq[(Workspace, Option[MultiregionalBucketMigrationProgress])]] =
+      workspaces
+        .traverse { workspace =>
+          getBucketLocation(workspace).map(workspace -> _)
+        }
+        .map {
+          _.collect {
+            case (workspace, Some(location)) if location.equals("US") =>
+              (workspace, Option.empty[MultiregionalBucketMigrationProgress])
+          }
+        }
+
+    def getMigratingOrRecentlyFinishedWorkspaces(
+      workspaces: Seq[(Workspace, Option[MultiregionalBucketMigrationProgress])]
+    ): Future[Seq[(Workspace, Option[MultiregionalBucketMigrationProgress])]] =
+      workspaces
+        .traverse {
+          case (workspace, progress @ Some(MultiregionalBucketMigrationProgress(_, Some(outcome), _, _))) =>
+            dataSource
+              .inTransaction(_.multiregionalBucketMigrationQuery.getMigrationAttempts(workspace))
+              .map { migrations =>
+                if (
+                  migrations.exists(
+                    _.finished
+                      .getOrElse(Timestamp.from(Instant.MIN))
+                      .after(Timestamp.from(Instant.now().minus(7, ChronoUnit.DAYS))) || outcome.isFailure
+                  )
+                ) Some(workspace -> progress)
+                else None
+              }
+          case unfinishedWorkspaceWithProgress @ (_, Some(MultiregionalBucketMigrationProgress(_, None, _, _))) =>
+            Future.successful(unfinishedWorkspaceWithProgress.some)
+          case _ =>
+            Future.successful(
+              None
+            )
+        }
+        .map(_.flatten)
+
+    for {
+      workspaces <- getOwnedWorkspaces
+
+      // get migration progress if it exists
+      workspacesWithProgressOpt <- workspaces.traverse { workspace =>
+        dataSource
+          .inTransaction(getBucketMigrationProgress(workspace))
+          .recover(_ => None)
+          .map(workspace -> _)
+      }
+
+      (scheduledWorkspaces, unscheduledWorkspaces) = workspacesWithProgressOpt.partition(_._2.isDefined)
+
+      eligibleWorkspaces <- getEligibleUnmigratedWorkspaces(unscheduledWorkspaces.map(_._1))
```

Comment:
Also this field.

Reply:
Actually, why doesn't `getEligibleUnmigratedWorkspaces` return only workspace? You don't use the empty  `Option.empty[MultiregionalBucketMigrationProgress]`, do you?

Reply:
I don't, it's just to make combining `eligibleWorkspaces` and `migratingOrRecentlyFinishedWorkspaces` in the response easier

Reply:
Oh, I misread the code. I thought somehow only the first part of the response from `getEligibleUnmigratedWorkspaces` was getting assigned to the variable `eligibleWorkspaces`. Would you mind naming it something more like `eligibleWorkspacesAndEmptyProgress` or …. I don't know, something that makes it clear that this is more than just a list of workspaces?

Reply:
Or perhaps the return types are enough of the helper methods. As usual, I simply find Scala difficult to read.

Reply:
:+1: I'll rename it -- it took me a bit to remind myself of what was happening there too

Reply:
Great, then I think we would need the other to be `migratingOrRecentlyFinishedWorkspacesAndProgress`.